proposals/NNNN-refining-property-wrapper-related-initialization.md

```diff
@@ -7,98 +7,198 @@
+First, we propose an update to the rules of the synthesized memberwise initializer for types with wrapped properties such that wrapper attributes are mapped directly into the initializer 
+For example, a type such as this one from TSPL:
+
+```swift
+struct MixedRectangle {
+    @SmallNumber var height: Int = 1
+    @SmallNumber(maximum: 9) var width: Int = 2
+}
+```
+
+would recieve a synthesized memberwise initializer that looks like:
+
+```swift
+init(@SmallNumber height: Int = 1, @SmallNumber(maximum: 9) width: Int = 2) {
+  ...
+}
+```
+
+Second, we propose allowing property wrapper storage for global, type, and local wrapped properties to be initialized via the `init(projectedValue:)` system (as discussed in the future directions of SE-0293). E.g.:
+
+```swift
+@Wrapper
+var property: Int
+
+$property = someProjectedValue
+```
 
 ## Detailed Design
 
-Describe the design of the solution in detail. If it involves new
-syntax in the language, show the additions and changes to the Swift
-grammar. If it's a new API, show the full API and its documentation
-comments detailing what it does. The detail in this section should be
-sufficient for someone who is *not* one of the authors to be able to
-reasonably implement the feature.
+### Synthesized memberwise initializer and property wrappers
+
+We propose a new algorithm for generating the synthesized memberwise initializer as it relates to wrapped properties. Rather than trying to decide whether to expose the storage type or the wrapped type in the initializer, a property declared in a type `MyType` as:
+
+```swift
+@Wrapper(arg: value)
+var property: X = initialValue
+```
+
+will always be represented in the synthesized memberwise initializer signature as:
+
+```swift
+init(..., @Wrapper(arg: value) property: X = initialValue, ...)
+```
+
+(where the argument list after `Wrapper` and the `initialValue` may or may not be present).
+
+Note that the rules for property wrappers in parameter lists all apply as normal. Most importantly:
+- If the property declaration _does_ provide arguments to the wrapper attribute, `Wrapper` will always be an implementation detail property wrapper.
+- If `Wrapper` provides an `init(projectedValue:)`, then the synthesized memberwise initializer will be callable via the `$` syntax, e.g.,
+    ```swift
+    let _ = MyType(..., $property: ProjectedValue(), ...)
+    ```
+    
+Refer to SE-0293 for a detailed discussion of how the property wrapper transformation proceeds for wrapped arguments.
+    
+If the specific initializers declared by the `Wrapper` type would result in an un-callable synthesized initializer (for example, because `Wrapper` provides no appropriate `init(wrappedValue:)` or `init(projectedValue:)`), an error will be emitted and the user will have to adjust their property declaration or define a custom initializer (see [**Source Compatibility**](#source-compatibility) below for more information).
```

Comment:
Hmm, in my mind the error was analogous to the error emitted on a class with no initializer:

```
class C { // error: class 'C' has no initializers
    var x: Int
}
```

Reply:
(Since this question could be raised in review, we should probably add a rationale for what we settle on.)

Good point. However, classes and structs differ in terms of subclassing and unsafe instantiation. 

Firstly, classes support subtyping. This means that the subclass must first initialize `super`, which is impossible (in safe code) with the `C` class above. (That's important because `C` also provides a stored property, potentially used by methods that'll be inherited by subclasses.) Admittedly, this is argument is undermined by the fact that adding `final` is still is erroneous; but, to be honest, this feels like an inadvertent design deficiency.

Another difference is the lack of straightforward unsafe instantiation. That is, we can unsafely get a `struct` instance without (accessible) initializers pretty easily. That's more difficult with classes, though, which are pointers to their reference count, payload, and other properties I'm unfamiliar with.

Perhaps I'm overthinking this, but I think an error is too limiting for structs when a warning would suffice. What's more, even if the user ignores the warning, it won't be long before they want to initialize their struct and encounter an actual error.

What do you think?